File: Baekjoon/Others/17281.py
# from sys import stdin; input = stdin.readline
from itertools import permutations

def main():
    # 0 input
    N = int(input())
    innings = [list(map(int, input().split())) for _ in range(N)]

    # 1 bruteforce
    max_score = 0
    # 1-1 make permutation
    for player_order_first in permutations(range(2, 10), 3):
        # 1-2 calculate max value / simulation
        other_players = set(range(2, 10)) - set(player_order_first)
        for player_order_second in permutations(other_players, 5):
            player_order = [*player_order_first, 1, *player_order_second]
            player_idx = 0
            score = 0
            for inning in innings:
                out = 0
                bases = [0 for _ in range(3)] # 1st ~ 3rd base
                while out < 3:
                    cur_player = player_order[player_idx] - 1
                    if 0 == inning[cur_player]:
                        out += 1
                    elif 1 == inning[cur_player]:
                        score += bases[2]
                        bases[0], bases[1], bases[2] = 1, bases[0], bases[1]
                    elif 2 == inning[cur_player]:
                        score += bases[1] + bases[2]
                        bases[0], bases[1], bases[2] = 0, 1, bases[0]
                    elif 3 == inning[cur_player]:
                        score += sum(bases)
                        bases[0], bases[1], bases[2] = 0, 0, 1
                    elif 4 == inning[cur_player]:
                        score += sum(bases) + 1
                        bases[0], bases[1], bases[2] = 0, 0, 0
                    player_idx = (player_idx + 1) % 9

            max_score = max(max_score, score)

    # 2 output
    print(max_score)

if __name__ == "__main__":
    main()

File: Baekjoon/Others/17281.cpp
#include <iostream>
#include <array>
#include <vector>
#include <algorithm> // next_permutation
using namespace std;
#define NUM_PLAYERS 9

bool play(int cur_player_play, int &score, array<int, 3> &bases)
{
  // cout << cur_player_play << " bases: ";
  // for (int i = 0; i < 3; i++)
  // {
  //   cout << bases[i] << " ";
  // }
  // cout << endl;

  // cout << "before score: " << score << endl;
  switch (cur_player_play)
  {
  case 0:
    return true;
  case 1:
    score += bases[2];
    bases[2] = bases[1];
    bases[1] = bases[0];
    bases[0] = 1;
    break;
  case 2:
    score += bases[1] + bases[2];
    bases[2] = bases[0];
    bases[1] = 1;
    bases[0] = 0;
    break;
  case 3:
    score += bases[0] + bases[1] + bases[2];
    bases[2] = 1;
    bases[1] = 0;
    bases[0] = 0;
    break;
  case 4:
    score += bases[0] + bases[1] + bases[2] + 1;
    bases[2] = 0;
    bases[1] = 0;
    bases[0] = 0;
    break;
  default:
    break;
  }

  // cout << "after score: " << score << endl;
  return false;
}

int main()
{
  // 0 initial setting
  ios::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);

  // 1 input
  int N;
  cin >> N;
  vector<array<int, 9>> innings;
  for (int i = 0; i < N; i++)
  {
    array<int, 9> inning;
    for (int j = 0; j < 9; j++)
    {
      cin >> inning[j];
    }
    innings.push_back(inning);
  }
  // 2 bruteforce
  int max_score = 0;
  // 2-1 make permutations
  array<int, 9> players_order = {1, 2, 3, 4, 5, 6, 7, 8, 9};
  int break_cnt = 0;
  do
  {
    if (players_order[3] != 1)
      continue;

    int player_idx = 0;
    int score = 0;
    for (int i = 0; i < N; ++i) // inning start, inning = innings[i]
    {
      int out = 0;
      array<int, 3> bases = {0, 0, 0};
      while (out < 3)
      {
        int cur_player = players_order[player_idx] - 1;
        if (play(innings[i][cur_player], score, bases))
        {
          out++;
        }
        player_idx = (player_idx + 1) % 9;
      }
    }
    if (max_score < score)
    {
      max_score = score;
    }

  } while (next_permutation(players_order.begin(), players_order.end()));

  // 3 output
  cout << max_score << '\n';

  return 0;
}